### ISSUE TITLE
Chapter 2: Overfitting 

### ISSUE THREAD

**Issue by laurent-pincemaille** [ageron/handson-ml#381]
Dear Sir,

I was wondering how you could find such a good result with a RandomForest Regressor compared to my own results; i finally understood that you let the max_depth parameter to its default value which is None.
I remember an interview with coding project  where i had let this parameter to None with along other values  (small one liike [2,4,8] and the interviewer told me: "never train a RF with the parameter max_depth set to None, because all your result are meaningless since  you overfit far too much".
So i was wondering how do you, in practice, try to avoid overfitting ? I personnally plot a validation Curve in order to vizualize the gap between the mean test and mean train score obtained from cross validation. If for one parameter the step/gap is "too hight" i overfit. But i have no idea of what should be the gap to say i overfit.

Thanks for your help

**Comment by ageron** [comment]
Hi @laurent-pincemaille ,

Thanks for your question. Indeed, I left the `max_depth` parameter set to `None` on purpose, to show that the model would completely overfit the training set. Here's an excerpt from that chapter:

> ```pycon
> >>> tree_mse = mean_squared_error(housing_labels, housing_predictions)
> >>> tree_rmse = np.sqrt(tree_mse)
> >>> tree_rmse
> 0.0
> ```
> 
> Wait, what!? No error at all? Could this model really be absolutely perfect? Of course, it is much more likely that the model has badly overfit the data. How can you be sure? As we saw earlier, you don't want to touch the test set until you are ready to launch a model you are confident about, so you need to use part of the training set for training, and part for model validation.
> 

The goal here was to show that a model should not be judged by its performance on the training set. That's why in the following section the book introduces cross-validation: it will give you a much more accurate measure of how well your model will perform once it's put in production. To determine how badly a model overfits, you just compare its performance on the training set and on the validation set (or using cross-validation). There's no clear rule to determine how bad overfitting is, it depends on the use case. For example, a good model may reach 95% accuracy on the training set, and just 94% on the validation set, and that's a bit of overfitting, but not too much (compare the error rates: 5% vs 6%). But if it's 98% accuracy on the training set, and 94% on the validation set, then the error rates are 2% vs 6%, that's a factor of 3, that's pretty severe, and you should certainly regularize your model (and/or choose a simpler model, and/or get more data).

To find the right level of regularization (e.g., using the `max_depth` hyperparameter), look at the hyper parameter tuning section: basically try out many different `max_depth` values, and keeping the one that performs has the lowest validation error (measured using a validation set, or using cross-validation).  By the way, in chapter 7 I listed several other hyperparameters you can tweak to regularize a hyperparameter, the `max_depth` is just one of them.

Hope this helps,
Aurélien
